### ISSUE TITLE
Implement caching of boxed values in ModelBase to decrease memory footprint

### ISSUE THREAD

**Issue by GeertvanHorrik** [Catel/Catel#1437]
We've tried this before (see #1319). Seems I am in a better mood this evening getting this to work.

**Comment by GeertvanHorrik** [comment]
It's not necessarily faster, but does have a better memory footprint, so will see if we can get all the tests to succeed with this one.

**Comment by vaecors** [comment]
So I looked into #1319 a bit and the behavior looks awfully close to something similar to what we have built but it uses precompiled Expression Trees. The setter is either an `Action<T>` (untyped) or `Action<T, TReturn>` (strongly typed) and the Getter is either a [Func<T>](url) (untyped) or `Func<T, TReturn>` (strongly typed). 

There are two higher level methods that can create all setters and getters for an input class (cached in an `IReadOnlyDictionary `with the PropertyName as a key stored as a case insensitive string). [FastExpressionCompiler](https://github.com/dadhi/FastExpressionCompiler) is used to speed up the compiling of the expressions but as it's only ran once the only expense is the cold start, you can reference from the dictionary going forward. 

We use this method for everything from projecting an IDataReader to a strongly typed object to dynamic properties.

Is this something that might interest you?

**Comment by GeertvanHorrik** [comment]
>So I looked into #1319 a bit and the behavior looks awfully close to something similar to what we have built but it uses precompiled Expression Trees. The setter is either an `Action<T>` (untyped) or `Action<T, TReturn>` (strongly typed) and the Getter is either a `Func<T>` (untyped) or `Func<T, TReturn>` (strongly typed).  ...

@vaecors absolutely!

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by vaecors** [comment]
@GeertvanHorrik I’m actually working on a PoC. I had a few ideas. When I’m confident with it will push up a repo.

**Comment by vaecors** [comment]
@GeertvanHorrik Alright so that PoC did not end up where I was hoping. However I have gone ahead and put up a [gist](https://gist.github.com/vaecors/fbb7bbafdf47b1e4e7e9b34ae9ab027a) of an expression tree builder utility class that does both untyped and typed setters and getters. You may be able to find some use for it. It does require the FastExpressionCompiler library as a dependency.

**Comment by GeertvanHorrik** [comment]
See #1438 for the PR implementing this change. Will see if I can use the expression stuff as well, but prefer to keep Catel.Core dependency-less.

**Comment by vaecors** [comment]
The FastExpressionCompiler is an optional behavioral change and is not necessarily required for those methods. We used it to speed up cold starts and the fact the library it sits in already depends on it due to our use of [Mapster](https://github.com/MapsterMapper/Mapster). I made about one comment concerning one of the changes. I also was a bit curious of the use of [IsValueType](https://docs.microsoft.com/en-us/dotnet/api/system.valuetype?view=netframework-4.8) vs [IsPrimitive ](https://docs.microsoft.com/en-us/dotnet/api/system.type.isprimitive?view=netframework-4.8) type for one of the calling methods but that's more of a preference I feel more than an advantage of A vs B.
